### ISSUE TITLE
Add feature for proxy definition

### ISSUE THREAD

**Issue by caalador** [vaadin/flow#7567]
Add feature and documentation on how to define proxies to be used with node/npm download and node/npm execution.

This could be a proxies file in ~/.vaadin or it could be that we would only use the .npmrc file for this.

For proxy data see [Proxy](https://github.com/vaadin/flow/blob/f121b3f21df85e8a29793653bb0f357c4a9c06b0/flow-server/src/main/java/com/vaadin/flow/server/frontend/copy/ProxyConfig.java#L58)

**Comment by mehdi-vaadin** [comment]
The documentation is added via https://github.com/vaadin/flow-and-components-documentation/pull/1108.

**Comment by knoobie** [comment]
@pleku Found the time to test this:

TL;DR - proxy settings are used, but there is no real information when the process fails

How did I test it:

1. Downloading starter (without anything) 14.1.16 
2. Updating Dependency and Repos to 14.2.0.alpha5
3. Starting squid proxy via docker (https://hub.docker.com/r/sameersbn/squid)
4. Creating a docker file to build

Dockerfile:
```
FROM maven:3.6.3-jdk-8

COPY .npmrc $HOME/.npmrc
```
.npmrc: (should be next to the docker file when creating)
```
proxy=172.17.0.1:3128
https-proxy=172.17.0.1:3128
```
5. Executing the build:

```
cd /your/path/to/the/project
docker run -v "$(pwd)":/usr/src/mymaven -v /etc/resolv.conf:/etc/resolv.conf --network=host -w /usr/src/mymaven tmp-npmrc-maven mvn -Pproduction clean install
```
FYI `tmp-npmrc-maven` is the name of the docker file

6. Expected behavior:

- the defined proxies are found (check!)
- if an error occurred while downloading (for example the proxy doesn't work) there a "helpful" message for the user

7. Current behavior:

When a proxy was found that doesn't allow connection (default squid) - there is no helpful error message about it like "Error while downloading Node"
Note: that comment about mvn com.github.eirslett looks outdated as well, or? Because you install it yourself?

```
[INFO] Couldn't find node. Installing Node and NPM to /root/.vaadin.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:55 min
[INFO] Finished at: 2020-02-26T21:16:40+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:14.2.0.alpha5:prepare-frontend (default) on project my-starter-project: 
[ERROR] 
[ERROR] ======================================================================================================
[ERROR] Vaadin requires node.js & npm to be installed. Please install the latest LTS version of node.js (with npm) either by:
[ERROR]   1) following the https://nodejs.org/en/download/ guide to install it globally. This is the recommended way.
[ERROR]   2) running the following Maven plugin goal to install it in this project:
[ERROR]   $ mvn com.github.eirslett:frontend-maven-plugin:1.7.6:install-node-and-npm -DnodeVersion="v12.14.0" 
[ERROR] 
[ERROR] Note that in case you don't install it globally, you'll need to install it again for another Vaadin project.
[ERROR] In case you have just installed node.js globally, it was not discovered, so you need to restart your system to get the path variables updated.
[ERROR] =========
```


FYI: When no proxy were found:
```
[INFO] Couldn't find node. Installing Node and NPM to /root/.vaadin.
[INFO] Installing node version v12.16.0
[INFO] Downloading https://nodejs.org/dist/v12.16.0/node-v12.16.0-linux-x64.tar.gz to /root/.vaadin/node-v12.16.0-linux-x64.tar.gz
[INFO] No proxies configured
[INFO] No proxy was configured, downloading directly
[INFO] Unpacking /root/.vaadin/node-v12.16.0-linux-x64.tar.gz into /root/.vaadin/node/tmp
[INFO] Copying node binary from /root/.vaadin/node/tmp/node-v12.16.0-linux-x64/bin/node to /root/.vaadin/node/node
[INFO] Extracting NPM
[INFO] Local node installation successful.
```

**Comment by knoobie** [comment]
Btw. I was wondering, coming from the maven / java-only world - why do I have to create a .npmrc file? Shouldn't it be enough when I configure my maven settings.xml with a proxy and vaadin reuses this config for:

1. Downloading node/npm if necessary 
2. Downloading npm packages (I know it's more challenging because it's no java / maven - but I would expect it as user)

Possible solution / ideas / brainstorming for 2)

- allow the user to configure a proxy explicitly inside the vaadin-maven-plugin (inside the pom) so you have all information you need (password etc)
- ^ this should be done for npm-registry as well
- vaadin-maven-plugin could use '--proxy' or other options directly
- vaadin-maven-plugin allows to configure a `npmArguments` where the developer can configure any additional properties he needs like (registry, proxy and so on), this arguments are added to the npm goal execution

Opinion? Personally I like the `npmArguments` option - it sounds "easy" to implement and it's common for java developer to configure something like that (Example: jaxb maven plugin)

**Comment by caalador** [comment]
I never got an exception on the download that would have given the download manually exception. I'll test setting up the same way you have it setup.

The reason to choose .npmrc is that then running npm will automatically get the proxysettings. It if was in maven settings for instance then we would need something extra for gradle and the ones running without the plugin (e.g. projects from start.spring.io)
The proxy is also taken from system properties and environment properties.

**Comment by caalador** [comment]
It's throwing a `IllegalArgumentException("Provided proxyUrl does not match the format protocol://user:password@server:port nor protocol://server:port")`, Sadly this seems to get cast as a ioexception and then eaten up resulting in the please install node exception stack trace.

But if you update the proxy to
```
proxy=http://172.17.0.1:3128
https-proxy=http://172.17.0.1:3128
```
It should work. I seem to get a Connection timeout with my setup at the moment. Sadly that is also eaten and not shown. I'll fix that so that the actual exceptions get thrown as is so that the user gets some relevant information on what failed.

**Comment by knoobie** [comment]
@caalador Thanks for reproducing! You were 10 seconds faster, restarted the process with -X.

If somebody needs the stacktrace:

```
    at com.vaadin.flow.plugin.maven.PrepareFrontendMojo.execute (PrepareFrontendMojo.java:143)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.IllegalStateException: 

======================================================================================================
Vaadin requires node.js & npm to be installed. Please install the latest LTS version of node.js (with npm) either by:
  1) following the https://nodejs.org/en/download/ guide to install it globally. This is the recommended way.
  2) running the following Maven plugin goal to install it in this project:
  $ mvn com.github.eirslett:frontend-maven-plugin:1.7.6:install-node-and-npm -DnodeVersion="v12.14.0" 

Note that in case you don't install it globally, you'll need to install it again for another Vaadin project.
In case you have just installed node.js globally, it was not discovered, so you need to restart your system to get the path variables updated.
======================================================================================================

    at com.vaadin.flow.server.frontend.FrontendUtils.getExecutable (FrontendUtils.java:701)
    at com.vaadin.flow.server.frontend.FrontendUtils.getNodeExecutable (FrontendUtils.java:357)
    at com.vaadin.flow.server.frontend.FrontendUtils.validateNodeAndNpmVersion (FrontendUtils.java:1020)
    at com.vaadin.flow.plugin.maven.PrepareFrontendMojo.execute (PrepareFrontendMojo.java:141)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

```

**Comment by caalador** [comment]
Created a PR to have the exception messages come through.
For different ways to set the proxy definitions could you @knoobie open a new issue with possible suggestions?

**Comment by knoobie** [comment]
Follow up ticket is created https://github.com/vaadin/flow/issues/7696
